### DISCUSSION TITLE
cannot send POST request

### DISCUSSION THREAD

**Discussion by Stevepurpose** [curl/curl#12379] [Q&A] [answered]
please why do I Keep getting this response whenever I try to send a POST request .please see below                                                                                              
curl -X POST -H "Content-Type : application/json" -d '{"firstName"Barry","lastName": "Sam","age":20,"classNumber":25,"address":"5 kings street"}' http://localhost:4500/api/students/
curl: (6) Could not resolve host: kings
curl: (3) unmatched close brace/bracket in URL position 7:
street}' http://localhost:4500/api/students/

**Comment by Stevepurpose** [comment]
So I noticed an error in my data, which I have corrected but still getting this                                                                                                                                curl -X POST -H "Content-Type : application/json" -d '{"firstName": "Barry",  "lastName": "Sam", "age": 20, "classNumber": 25, "address": "maribor street"}' http://localhost:4500/api/students/                                                                                                                                                                curl: (3) URL using bad/illegal format or missing URL
curl: (3) URL using bad/illegal format or missing URL
curl: (3) URL using bad/illegal format or missing URL
curl: (3) URL using bad/illegal format or missing URL
curl: (3) URL using bad/illegal format or missing URL
curl: (3) URL using bad/illegal format or missing URL
curl: (3) URL using bad/illegal format or missing URL
curl: (3) URL using bad/illegal format or missing URL
curl: (3) unmatched close brace/bracket in URL position 15:
maribor street}'

**Comment by jay** [comment, marked as answer]
> curl: (3) unmatched close brace/bracket in URL position 7:

It looks like you are in Windows using the command shell (command prompt). In that case single-quoting is [not available](https://curl.se/docs/faq.html#Why_do_my_single_double_quotes_f) and you have to use double quotes to encapsulate the argument. Inner double quotes will need to be escaped. For example:

~~~
-d "{\"firstName\": \"Barry\", \"lastName\": \"Sam\", \"age\": 20, \"classNumber\": 25, \"address\": \"maribor street\"}"
~~~

`-X POST` is [unnecessary](https://daniel.haxx.se/blog/2015/09/11/unnecessary-use-of-curl-x/) and often wrong.

`-H "Content-Type : application/json" -d ...` can often be replaced by [--json](https://curl.se/docs/manpage.html#--json) if you are expecting a json reply. For example:

~~~
--json "{\"firstName\": \"Barry\", \"lastName\": \"Sam\", \"age\": 20, \"classNumber\": 25, \"address\": \"maribor street\"}"
~~~

Powershell (the other popular Windows shell) you can use single quotes for the argument but it's deceiving because when it passes the argument to a Windows executable (like curl) it will convert them to double quotes. So, in either case you have to escape the inner double quotes.

You can also put the json in a file and then tell curl to read the contents and send it as json:

~~~
--json @filename
~~~

**Comment by Stevepurpose** [comment]
Thanks a lot for your help and reply. I know about the alternatives. I am
trying to prepare a lecture so I am trying to get my facts straight. I
think your reply did a good summary
